src/profiles/profiles.rs

```diff
@@ -0,0 +1,519 @@
+use std::{collections::BTreeMap, env, ffi::OsString, fs::read_to_string, path::PathBuf};
+
+use anyhow::{bail, Result};
+use log::{debug, info, warn};
+use semver::Version;
```

Comment:
I am using it. semver is being used for the Profile API version

Reply:
If it's just a number  it means we don't keep backwards compatibility?

Reply:
My goal was to leave us space to define API changes that would not break it so future versions of the tool could define a VersionReq and have proper semver checks.

Reply:
In fact, we could very much add `VersionReq("1") right now.

Reply:
Early bailing is probably not going to be a consistent feature unless we do two-phase parsing like with k8s objects, eg.:
```yaml
type:  Profile
version: 1.2.3
spec:
   - (actual object)
```

I sill think two numbers are needed. A profile version X.Y means a retis version that supports at least X.0 should be able to parse and apply a reasonable profile. It might miss stuff in which case we should warn/inform that the full configuration might be missing but we continue running.

A retis version that does not support at least X.0 can:

- Fail to parse the profile. In this case retis will dump it's supported version and fail to start
- Properly parse the profile, read the version and realize that although it might understand it, the missing parts make the configuration inconsistent and should not attempt to continue. In this case, we should fail gracefully.

Reply:
My point is that if if we increase `X` is a way of encoding that, _even if the profile is parseable_ it should not be processed as the missing parts would be too important or it would make the tool fail further down the road in a more difficult to understand way. Where as increasing `Y` means that, if parseable, a retis with matching `X` should try to apply and warn.

If we only use one number, we would have to maintain a compatibility matrix in retis. For each i < N we would have to know code if the version is apply-able or not. I think that's too complex to maintain. If we go for a single number I would make it more strict, i.e: fail to load any version different to current.

Reply:
My point is that if if we increase `X` is a way of encoding that, _even if the profile is parseable_ it should not be processed as the missing parts would be too important or it would make the tool fail further down the road in a more difficult to understand way. Where as increasing `Y` means that, if parseable, a retis with matching `X` should try to apply and warn.

If we only use one number, we would have to maintain a compatibility matrix in retis. For each i < N we would have to know code if the version is apply-able or not. I think that's too complex to maintain. If we go for a single number I would make it more strict, i.e: fail to load any version different to current.

Reply:
Exactly, the only difference is where the logic is. Whether defined in the version itself and a well-known semantics or coded inside the application.

In order to avoid an always increasing compatibility matrix inside, I'd say if go for single value we should really restrict compatibility to the bare minimum. 

Also related to this (and probably helpful to agree o the same versioning strategy) is what we consider an API change. On the one end, we can consider each option that's configurable via profile a part of the API, so adding a collector or an option to one would be an API change (in the sense that a profile that uses it will make a retis that doesn't fail). Alternatively, we could decide to take the content of "args:" object out of the API, in which case we're only considering the top level Proflie schema and actually applying the configuration is best effort.

Thoughts?

Reply:
If we choose the former, `Version` can be thought as a condition (same as kernel version and symbol existence) over `retis` itself. If we choose the latter, then I think a single number (or even removing the version and making unmarshalling more strict) will work.

Reply:
> But I'm not sure I fully got the above, do you propose to use this selector in addition to the API version, or instead of?

I was thinking maybe it can be used instead of.

Reply:
I think you're right. We should have different fields. Let's have an "api_version" for now and work on parallel on a retis version condition